File: UVA/AOAPC-BAC/Ch9/Example/UVA-116_Unidirectional TSP.cpp
#include <bits/stdc++.h>
using namespace std;
const int INF = 0x3f3f3f3f;
#define endl '\n'

vector<int> DX = {-1, 0, 1};
int n, m;
vector<vector<int>> grid, memo, nxt;

int dfs(int x, int y) {
    if (y == m - 1) return grid[x][y];
    int &res = memo[x][y];
    if (res != INF) return res;
    for (int dx : DX) {
        int nx = (x + dx + n) % n; // 環狀處理
        int nd = dfs(nx, y + 1) + grid[x][y];
        if (nd < res || (nd == res && nx < nxt[x][y])) {
            res = nd;
            nxt[x][y] = nx; // 記錄下一個 column 是哪一個 row
        }
    }
    return res;
}

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    while (cin >> n >> m) {
        grid.assign(n, vector<int>(m));
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                cin >> grid[i][j];
            }
        }
        // dfs(x, y) 表示從 (x, y) 到最後一個 column 的最短路徑
        memo.assign(n, vector<int>(m, INF));
        nxt.assign(n, vector<int>(m, -1)); // 紀錄路徑

        // 求從第一個 column 的任意位置到最後一個 column 的最短路徑
        int ans = INF, idx = 0;
        for (int i = 0; i < n; i++) {
            int res = dfs(i, 0);
            if (res < ans) ans = res, idx = i;
        }
        // 輸出路徑
        for (int y = 0; y < m; y++) {
            cout << idx + 1 << (y == m - 1 ? '\n' : ' ');
            idx = nxt[idx][y];
        }
        cout << ans << endl;
    }
    return 0;
}

File: UVA/AOAPC-BAC/Ch9/Example/UVA-116_Unidirectional TSP.py
from functools import cache

import sys
buf = sys.stdin.read().split()
cin = lambda: buf.pop(0)

while True:
    try:
        n, m = int(cin()), int(cin())
    except IndexError:
        break
    grid = [[int(cin()) for _ in range(m)] for _ in range(n)]

    next = [[-1] * m for _ in range(n)] # 紀錄路徑
    # dfs(i, j) 表示從 (i, j) 到終點的最短路徑
    @cache
    def dfs(i, j):
        if j == m - 1:
            return grid[i][j]
        res = float('inf')
        for di in [-1, 0, 1]:
            ni = (i + di) % n
            nd = dfs(ni, j + 1) + grid[i][j]
            if nd < res or (nd == res and ni < next[i][j]):
                res = nd
                next[i][j] = ni
        return res

    ans = dfs(0, 0)
    i = 0
    for j in range(m):
        print(i + 1, end=' ' if j < m - 1 else '\n')
        i = next[i][j]
    print(ans)